File: programmers/합승 택시 요금/합승 택시 요금.cpp
#include <string>
#include <vector>
using namespace std;

int LIMITS = 30000000;
int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    int answer = 0;
    
    vector<vector<int>> node(n + 1);
    for (int i = 1; i <= n; i++) {
        node[i].assign(n + 1, LIMITS);
    }
    
    for (auto fare : fares) {
        node[fare[0]][fare[1]] = fare[2];
        node[fare[1]][fare[0]] = fare[2];
    }
    
    for (int check = 1; check <= n; check++) {
        for (int x = 1; x <= n; x++) {
            if (x == check) continue;
            for (int y = 1; y <= n; y++) {
                if (y == check) continue;
                int distance = node[x][check] + node[check][y];
                if (distance < node[x][y]) {
                    node[x][y] = distance;
                }
            }
        }
    }
    
    answer = node[s][a] + node[s][b];
    for (int check = 1; check <= n; check++) {
        int course = node[s][check];
        if (a != check) course += node[check][a];
        if (b != check) course += node[check][b];
        answer = min(answer, course);
    }
    return answer;
}

File: programmers/합승 택시 요금/합승 택시 요금.kt
fun solution(n: Int, s: Int, a: Int, b: Int, fares: Array<IntArray>): Int {
    val limit = 30000000
    val taxi = Array(n) { IntArray(n) { limit } }
    
    repeat(n) {
        taxi[it][it] = 0
    }
    fares.forEach {
        val c = it[0] - 1
        val d = it[1] - 1
        taxi[c][d] = it[2]
        taxi[d][c] = it[2]
    }
    repeat(n) { check ->
        repeat(n) start@ { x ->
            if (x == check) return@start
            repeat(n) end@ { y ->
                if ((y == check) or (x == y)) return@end
                val distance = taxi[x][check] + taxi[check][y]
                taxi[x][y] = Math.min(taxi[x][y], distance)
            }
        }
    }
    var answer = taxi[s - 1][a - 1] + taxi[s - 1][b - 1]
    repeat(n) { check -> 
        answer = Math.min(answer, taxi[s - 1][check] + taxi[check][a - 1] + taxi[check][b - 1])
    }
    return answer
}